File: teamcenter/src/main/resources/definitions/TeamCenter/Create.js
/*
 * Copyright 2006 The eFaps Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author:          tmo
 * Revision:        $Rev$
 * Last Changed:    $Date$
 * Last Changed By: $Author$
 */

/**
 * The private function creates all Team Center tables.
 *
 * @param _stmt SQL statement to work on
 */
function _eFapsCreateAllTeamCenterTables(_con, _stmt)  {

  print("Create TeamCenter Tables");

  eFapsCommonSQLTableCreate(_con, _stmt, "Folder object", "TCFOLDER", "ABSTRACT",[
      ["PARENTFOLDER          "+TYPE_INTEGER],
      ["constraint TCFOLDER_FK_PRNTFOL foreign key(PARENTFOLDER) references TCFOLDER(ID)"]
  ]);

  eFapsCommonSQLTableCreate(_con, _stmt, "Document object", "TCDOCUMENT", "ABSTRACT",[
      ["FILENAME              "+TYPE_STRING_SHORT+"(128)"],
      ["FILELENGTH            "+TYPE_INTEGER]
  ]);

  eFapsCommonSQLTableCreate(_con, _stmt, "Connection beetween Documents and Folders", "TCDOC2FOL", null,[
      ["TCDOCUMENT            "+TYPE_INTEGER+"                   not null"],
      ["TCFOLDER              "+TYPE_INTEGER+"                   not null"],
      ["CREATOR               "+TYPE_INTEGER+"                   not null"],
      ["CREATED               "+TYPE_DATETIME+"                  not null"],
      ["MODIFIER              "+TYPE_INTEGER+"                   not null"],
      ["MODIFIED              "+TYPE_DATETIME+"                  not null"],
      ["constraint TCDOC2FOL_UK_FOLDOC unique(TCDOCUMENT,TCFOLDER)"],
      ["constraint TCDOC2FOL_FK_TCDOC  foreign key(TCDOCUMENT)   references ABSTRACT(ID)"],
      ["constraint TCDOC2FOL_FK_TCFLDR foreign key(TCFOLDER)     references TCFOLDER(ID)"],
      ["constraint TCDOC2FOL_FK_CRTR   foreign key(CREATOR)      references USERPERSON(ID)"],
      ["constraint TCDOC2FOL_FK_MDFR   foreign key(MODIFIER)     references USERPERSON(ID)"]
  ]);
}


//_eFapsCreateAllTeamCenterTables(con, stmt);

File: kernel-install/src/main/resources/org/efaps/kernel-install/definitions/Common/UIForm/ESJP/org/efaps/esjp/common/uiform/Create.java
/*
 * Copyright 2003-2007 The eFaps Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Revision:        $Rev$
 * Last Changed:    $Date$
 * Last Changed By: $Author$
 */

package org.efaps.esjp.common.uiform;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import org.efaps.admin.datamodel.Attribute;
import org.efaps.admin.datamodel.attributetype.AbstractFileType;
import org.efaps.admin.event.EventExecution;
import org.efaps.admin.event.Parameter;
import org.efaps.admin.event.Return;
import org.efaps.admin.event.Parameter.ParameterValues;
import org.efaps.admin.ui.CommandAbstract;
import org.efaps.admin.ui.Field;
import org.efaps.db.Context;
import org.efaps.db.Insert;
import org.efaps.db.Instance;
import org.efaps.util.EFapsException;

/**
 * @author jmo
 * @version $Id$
 * @todo description
 */
public class Create implements EventExecution {
  /**
   * Logger for this class
   */
  private static final Log LOG = LogFactory.getLog(Create.class);

  public Return execute(final Parameter _parameter) {
    Return ret = new Return();
    Instance instance = (Instance) _parameter.get(ParameterValues.INSTANCE);
    CommandAbstract command =
        (CommandAbstract) _parameter.get(ParameterValues.UIOBJECT);
    try {
      Context context = Context.getThreadContext();

      Insert insert = new Insert(command.getTargetCreateType());
      for (Field field : command.getTargetForm().getFields()) {
        if (field.getExpression() != null
            && (field.isCreatable() || field.isHidden())) {
          Attribute attr =
              command.getTargetCreateType().getAttribute(field.getExpression());
          if (attr != null
              && !AbstractFileType.class.isAssignableFrom(attr
                  .getAttributeType().getClassRepr())) {
            String value = context.getParameter(field.getName());
            insert.add(attr, value);
          }
        }
      }
      if (command.getTargetConnectAttribute() != null) {

        insert.add(command.getTargetConnectAttribute(), "" + instance.getId());
      }
      insert.execute();
    } catch (EFapsException e) {
      LOG.error("execute(Parameter)", e);
    }
    return ret;
  }
}
